src/third/redux-persist/createPersistor.js

```diff
@@ -35,19 +35,20 @@ export default function createPersistor (store, config) {
   let writeInProgress = false
 
   store.subscribe(() => {
-    if (paused || writeInProgress) return
-    writeInProgress = true
+    if (paused || writeInProgress) return;
 
-    const state = store.getState()
-    const updatedSubstates = []
+    (async () => {
+      writeInProgress = true
 
-    Object.keys(state).forEach((key) => {
-      if (!passWhitelistBlacklist(key)) return
-      if (lastState[key] === state[key]) return
-      updatedSubstates.push([key, state[key]])
-    });
+      const state = store.getState()
```

Comment:
Moving this inside the async IIFE actually represents a change that I think is worth noting in the commit message: it means we have a yield before reading what the new state is that we're going to be writing out.

I think that's fine, and indeed probably an improvement whenever it makes a difference. If there's a burst of updates that happen synchronously (e.g., if anywhere in the code we dispatch one plain action and then another, with no `await` in between), then the code before this commit would fire this callback on the first update; set `writeInProgress` and read the state; then on the second update, do nothing; and have a task come back later to start writing, and it'd write out the first updated state even though the second one exists by then. With this change, we wait to read the new state until that task gets run, which means we take the latest state that's then available and we write that.

Reply:
> it means we have a yield before reading what the new state is that we're going to be writing out.

Interesting. I don't think I follow, yet, though. To find the yield you're talking about, I'm looking for an `await` in the async IIFE's body, before the `const state = store.getState()` line. I don't see one, so I don't see the yield.

Is there a hidden yield that gets done without `await`, maybe at the start of invoking any async function?

Reply:
Hmm, no, that's right -- I think what I wrote above is mistaken.

This kind of question can be confusing (evidently!), in part because it can vary between languages. Here's notes I took on it a few years ago, with an empirical demo:
https://github.com/gnprice/async-playground/blob/7da42982af9e05540833a96ef7cb82dbaa397b92/async.js

Those confirm that there's no yield at entry to an async function. There is at exit (i.e., when the caller awaits the result), as well as at any awaits inside its body.

I also appear to have been confused about this when I wrote this other comment:
https://github.com/zulip/zulip-mobile/pull/4694#discussion_r691685866
> In the old code, we invoke the async IIFE, then update `lastState` and return. Then later when the whole JS stack is empty, the task created by that async IIFE gets run, and the code in its body starts executing. If the loop condition is true so it hits the `await` inside there, it creates a new task and returns […]

This point doesn't change much there, though: it basically makes the sequence a step shorter.